Dynect fog

ayashjorden opened the pull request:
Rewrite DynECT provider to support multiple contnet items for certain record name, e.g. different IPs pointing the same address, thus utilizing DNS round-robin ability.

Commit:
Make DynECT provider use Fog

Commit:
Bump ruby verison to 2.3.1 in .travis.yml

ayashjorden commented:
Hi @charlesdunbar,
Can I ask you to take a look at this?

Thank you,
Yarden

charlesdunbar commented:
Thanks for this PR!  Switching to fog has been on my back burner for a while - will take a look at this sometime this week.

charlesdunbar commented:
The main issue I'm seeing with this is the removal of self.prefetch.  As discussed in https://github.com/puppetlabs/puppetlabs-dns/pull/4#discussion_r28732262, without it, for every record that's managed you'll be grabbing the zone file over and over again.

charlesdunbar commented on README.md:
```diff
@@ -80,6 +80,16 @@ dns_record { "test-1a-record.puppetware.org":
+# content can also accept array
+
+dns_record { "test-1a-record.puppetware.org":
+  ensure  => present
```
Needs a trailing comma to be valid puppet code - which I'm now seeing is also the case with my examples elsewhere in the README.

ayashjorden replied:
FIxed that

ayashjorden commented:
I understand, something like [this](https://github.com/puppetlabs/puppetlabs-chronos/blob/master/lib/puppet/provider/chronos_job/chronos_job.rb#L86) ?

I need to read more about prefetch....

Commit:
Format README.md

Validate Puppet code
Make 'ensure' be first in resource attributes

charlesdunbar commented:
Slightly - We can't really use "instances" since for this case, instances would be what Dyn sees, which we can't check because by the time you run "instances" - you don't have the credentials available to pass to Dyn.

http://garylarizza.com/blog/2013/12/15/seriously-what-is-this-provider-doing/ has some good information about reading up on prefetch.

I'll probably have to go through this agains since I haven't looked at this code in about a year, but I highly recommend adding `require 'ruby-debug'; debugger` in-line to see the prefetch method in action on the master branch.  You're able to go line-by-line when you hit the debugger line, and print any variables at the time.

I should have some time tonight to refresh myself on the code and see if I can help flesh out a new prefetch.

Commit:
Query DNS zone with relevant node-string (based on the resource[:name])

No need to retrieve all of the DNS records and then filter what we need

ayashjorden commented:
Hi @charlesdunbar ,
I've read the link you provided about prefetch and also additional articles I've found on the web and things aren't clear yet.

Another somewhat related issue is that fog-dynect gem is missing zone and records TTL attribute when pulling data from DynECT. I've submitted a [PR](https://github.com/fog/fog-dynect/pull/12), hope to get it released soon.

Regarding the `flush` method implementation, I should call `r.save` only once as it `publish`s the DNS zone applyin TTL changes to *all* records under a node (node.name.domain.com.)

Commit:
Comment and format record comparison

Commit:
Fix TTL updating

Remove type comparison as we're alredy filtering when we query the zone for the
domain.
Call record.save only once as it publishes the DNS zone.

Per DNS RFC, all records under the same node must have the same TTL

ayashjorden committed:
Retire rest-client gem, not in use after the DynECT provider migration to 'fog'

ayashjorden commented:
Hi @charlesdunbar,
The fog-dynect gem Accepted my PR and released [v0.2.0](https://github.com/fog/fog-dynect/tree/v0.2.0) :+1: 

I've also fixed the issue with calling `save` only once.

We're left with the `prefetch` implementation issue.
I've read the code from _master_ and it fetches each record's state from DynECT. the _instances_ method does nothing.

On _dynect_fog_ branch, `instances` method just creates the resources' instances. Sync-ing the resources' state is done in the flush method.

Can you elaborate on the desired role of the `prefetch` method in the _dynect_fog_ context?

Thank you,
Yarden

ayashjorden committed:
Handle missing FQDN

Rescue Excon::Error::NotFound when FQDN doesn't exist

ayashjorden commented:
Hi @charlesdunbar,
Any update regarding `prefetch`?

ayashjorden commented:
Hi again @charlesdunbar,
On IRC, Dominic explained:
> [12:54:09] <Dominic> yardenbar: data about every instance of the resource already on the node (e.g. every package currently installed, every service configured)

I'll work on it :)
Appreciate your comment...

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-dns?pullRequest=7) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-dns?pullRequest=7) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Yarden Bar<br/>:x: ayashjorden<br/><hr/>**Yarden Bar** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-dns?pullRequest=7) it.</sub>
